File: simdify/simd.zig
const std = @import("std");
const v = @import("vector_types.zig");

// Shift packed 16-bit integers in a right by imm8 while shifting in zeros,
// and store the results in dst.
pub extern fn mm_srli_epi16(a: v.u8x16, b: u32) v.u8x16;
pub inline fn rshift16_m128(a: v.u8x16, b: u32) v.u8x16 {
    return mm_srli_epi16(a, b);
}

pub extern fn mm256_movemask_epi8(a: v.u8x32) u32;
pub inline fn movemask8_m256(a: v.u8x32) u32 {
    return mm256_movemask_epi8(a);
}

pub extern fn mm256_cmpeq_epi16(a: v.u8x32, b: v.u8x32) v.u8x32;
pub inline fn cmpeq16_m256(a: v.u8x32, b: v.u8x32) v.u8x32 {
    return mm256_cmpeq_epi16(a, b);
}

pub extern fn mm256_cmpeq_epi8(a: v.u8x32, b: v.u8x32) v.u8x32;
pub inline fn cmpeq8_m256(a: v.u8x32, b: v.u8x32) v.u8x32 {
    return mm256_cmpeq_epi8(a, b);
}

pub extern fn mm256_set_epi32(i0: u32, i1: u32, i2: u32, i3: u32, i4: u32, i5: u32, i6: u32, i7: u32) v.u8x32;

pub extern fn mm256_set_epi16(v15: u16, v14: u16, v13: u16, v12: u16, v11: u16, v10: u16, v9: u16, v8: u16, v7: u16, v6: u16, v5: u16, v4: u16, v3: u16, v2: u16, v1: u16, v0: u16) v.u8x32;
pub inline fn set16_m256(v15: u16, v14: u16, v13: u16, v12: u16, v11: u16, v10: u16, v9: u16, v8: u16, v7: u16, v6: u16, v5: u16, v4: u16, v3: u16, v2: u16, v1: u16, v0: u16) v.u8x32 {
    return mm256_set_epi16(v15, v14, v13, v12, v11, v10, v9, v8, v7, v6, v5, v4, v3, v2, v1, v0);
}
pub inline fn setall16_m256(a: u16) v.u8x32 {
    return mm256_set_epi16(a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a);
}

pub extern fn mm256_set_epi8(v32: u8, v30: u8, v29: u8, v28: u8, v27: u8, v26: u8, v25: u8, v24: u8, v23: u8, v22: u8, v21: u8, v20: u8, v19: u8, v18: u8, v17: u8, v16: u8, v15: u8, v14: u8, v13: u8, v12: u8, v11: u8, v10: u8, v9: u8, v8: u8, v7: u8, v6: u8, v5: u8, v4: u8, v3: u8, v2: u8, v1: u8, v0: u8) v.u8x32;
pub inline fn set8_m256(v32: u8, v30: u8, v29: u8, v28: u8, v27: u8, v26: u8, v25: u8, v24: u8, v23: u8, v22: u8, v21: u8, v20: u8, v19: u8, v18: u8, v17: u8, v16: u8, v15: u8, v14: u8, v13: u8, v12: u8, v11: u8, v10: u8, v9: u8, v8: u8, v7: u8, v6: u8, v5: u8, v4: u8, v3: u8, v2: u8, v1: u8, v0: u8) v.u8x32 {
    return mm256_set_epi8(v32, v30, v29, v28, v27, v26, v25, v24, v23, v22, v21, v20, v19, v18, v17, v16, v15, v14, v13, v12, v11, v10, v9, v8, v7, v6, v5, v4, v3, v2, v1, v0);
}
pub inline fn setall8_m256(a: u8) v.u8x32 {
    return mm256_set_epi16(a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a, a);
}

// Shuffle packed 8-bit integers in a according to shuffle control mask
// in the corresponding 8-bit element of b, and store the results in dst.
//
// Lệnh này tra cứu đồng thời 16-byte register (or lane)
// sử dụng 4-bit indices từ một vector khác:
//
// for (int i=0; i < VECTOR_SIZE; i++) {
//     uint8_t index = indices_vector[i];
//     if (index & 0x80) // 0x80: 0b10000000
//         result[i] = 0x00;
//     else
//         result[i] = lookup_vector[index & 0x0f]; // 0x0f = 0b1111
// }
//
// Trả về lookup_vector[n], với n = 4 bit thấp nhất của index
// Nếu bit thứ 8 của index = 1 thì kết quả luôn là 0 (clear flag)
// pshufb: Packed Shuffle Bytes (felixcloutier.com/x86/pshufb.html)
pub extern fn mm_shuffle_epi8(a: v.u8x16, b: v.u8x16) v.u8x16;
pub inline fn pshufb_m128(a: v.u8x16, b: v.u8x16) v.u8x16 {
    return mm_shuffle_epi8(a, b);
}

// Set packed 8-bit integers in dst with the supplied values in reverse order.
// dst[7:0] := v15, dst[15:8] := v14 .. dst[127:120] := v0
// dst = destination for shot
pub extern fn mm_setr_epi8(v15: u8, v14: u8, v13: u8, v12: u8, v11: u8, v10: u8, v9: u8, v8: u8, v7: u8, v6: u8, v5: u8, v4: u8, v3: u8, v2: u8, v1: u8, v0: u8) v.u8x16;
pub inline fn setrev8_m128(v15: u8, v14: u8, v13: u8, v12: u8, v11: u8, v10: u8, v9: u8, v8: u8, v7: u8, v6: u8, v5: u8, v4: u8, v3: u8, v2: u8, v1: u8, v0: u8) v.u8x16 {
    return mm_setr_epi8(v15, v14, v13, v12, v11, v10, v9, v8, v7, v6, v5, v4, v3, v2, v1, v0);
}

// Compare packed 8-bit integers in a and b for equality, and store the results in dst.
pub extern fn mm_cmpeq_epi8(a: v.u8x16, b: v.u8x16) v.u8x16;
pub inline fn cmpeq8_m128(a: v.u8x16, b: v.u8x16) v.u8x16 {
    return mm_cmpeq_epi8(a, b);
}

// Compare packed signed 8-bit integers in a and b for less-than, and store the results in dst.
// Note: This intrinsic emits the pcmpgtb instruction with the order of the operands switched.
pub extern fn mm_cmplt_epi8(a: v.u8x16, b: v.u8x16) v.u8x16;
pub inline fn cmplt8_m128(a: v.u8x16, b: v.u8x16) v.u8x16 {
    return mm_cmplt_epi8(a, b);
}

// Blend packed 8-bit integers from a and b using mask, and store the results in dst.
// FOR j := 0 to 15
//     i := j*8
//     IF mask[i+7]
//         dst[i+7:i] := b[i+7:i]
//     ELSE
//         dst[i+7:i] := a[i+7:i]
//     FI
// ENDFOR
pub extern fn mm_blendv_epi8(a: v.u8x16, b: v.u8x16, c: v.u8x16) v.u8x16;
pub inline fn blend8_m128(a: v.u8x16, b: v.u8x16, c: v.u8x16) v.u8x16 {
    return mm_blendv_epi8(a, b, c);
}

// Broadcast 8-bit integer a to all elements of dst. This intrinsic may generate vpbroadcastb.
pub extern fn mm_set1_epi8(x: u8) v.u8x16;
pub inline fn setall8_m128(x: u8) v.u8x16 {
    return mm_set1_epi8(x);
}

// Compute the bitwise AND of 128 bits (representing integer data) in a and b,
// and store the result in dst.
pub extern fn mm_and_si128(a: v.u8x16, b: v.u8x16) v.u8x16;
pub inline fn and_m128(a: v.u8x16, b: v.u8x16) v.u8x16 {
    return mm_and_si128(a, b);
}

pub extern fn mm_xor_si128(a: v.u8x16, b: v.u8x16) v.u8x16;
pub inline fn xor_m128(a: v.u8x16, b: v.u8x16) v.u8x16 {
    return mm_xor_si128(a, b);
}

pub extern fn mm_or_si128(a: v.u8x16, b: v.u8x16) v.u8x16;
pub inline fn or_m128(a: v.u8x16, b: v.u8x16) v.u8x16 {
    return mm_or_si128(a, b);
}

File: simdify/simd.c
#include <stdint.h>
#include <emmintrin.h> // SSE2 intrinsics 126-bit register
#include <immintrin.h> // Intel intrinsics, including 256-bit lane

__m256i mm256_shuffle_epi8(__m256i a, __m256i b) {
    return _mm256_shuffle_epi8(a, b);
}

uint32_t mm256_movemask_epi8(__m256i a) {
    return _mm256_movemask_epi8(a);
}

__m256i mm256_cmpeq_epi16(__m256i a, __m256i b) {
    return _mm256_cmpeq_epi16(a, b);
}

__m256i mm256_cmpeq_epi8(__m256i a, __m256i b) {
    return _mm256_cmpeq_epi8(a, b);
}

__m256i mm256_set_epi32(int i0, int i1, int i2, int i3,
                  int i4, int i5, int i6, int i7) {
  return _mm256_set_epi32(i0, i1, i2, i3, i4, i5, i6, i7);
}

__m256i mm256_set_epi16(short __w15, short __w14, short __w13, short __w12,
                 short __w11, short __w10, short __w09, short __w08,
                 short __w07, short __w06, short __w05, short __w04,
                 short __w03, short __w02, short __w01, short __w00) {
    return _mm256_set_epi16(__w15, __w14, __w13, __w12,
                 __w11, __w10, __w09, __w08,
                 __w07, __w06, __w05, __w04,
                 __w03, __w02, __w01, __w00);
}

__m256i mm256_set_epi8(char __b31, char __b30, char __b29, char __b28,
                 char __b27, char __b26, char __b25, char __b24,
                 char __b23, char __b22, char __b21, char __b20,
                 char __b19, char __b18, char __b17, char __b16,
                 char __b15, char __b14, char __b13, char __b12,
                 char __b11, char __b10, char __b09, char __b08,
                 char __b07, char __b06, char __b05, char __b04,
                 char __b03, char __b02, char __b01, char __b00) {
  return _mm256_set_epi8(__b31, __b30, __b29, __b28,
                 __b27, __b26, __b25, __b24,
                 __b23, __b22, __b21, __b20,
                 __b19, __b18, __b17, __b16,
                 __b15, __b14, __b13, __b12,
                 __b11, __b10, __b09, __b08,
                 __b07, __b06, __b05, __b04,
                 __b03, __b02, __b01, __b00);
}


__m128i mm_cmpeq_epi8(__m128i a, __m128i b) {
    return _mm_cmpeq_epi8(a, b);
}
__m128i mm_cmplt_epi8(__m128i a, __m128i b) {
    return _mm_cmplt_epi8(a, b);
}
__m128i mm_blendv_epi8(__m128i a, __m128i b, __m128i c) {
    return _mm_blendv_epi8(a, b, c);
}
__m128i mm_set1_epi8(signed char v) {
    return _mm_set1_epi8(v);
}

__m128i mm_setr_epi8(
    signed char v15,
    signed char v14,
    signed char v13,
    signed char v12,
    signed char v11,
    signed char v10,
    signed char v9,
    signed char v8,
    signed char v7,
    signed char v6,
    signed char v5,
    signed char v4,
    signed char v3,
    signed char v2,
    signed char v1,
    signed char v0) {
    return _mm_setr_epi8(v15,v14,v13,v12,v11,v10,v9,v8,
                          v7, v6, v5, v4, v3, v2,v1,v0);
}

__m128i mm_srli_epi16(__m128i a, signed int b) {
    return _mm_srli_epi16(a, b);
}

__m128i mm_and_si128(__m128i a, __m128i b) {
    return _mm_and_si128(a, b);
}


__m128i mm_or_si128(__m128i a, __m128i b) {
    return _mm_or_si128(a, b);
}

__m128i mm_xor_si128(__m128i a, __m128i b) {
    return _mm_xor_si128(a, b);
}

__m128i mm_shuffle_epi8(__m128i a, __m128i b) {
    return _mm_shuffle_epi8(a, b);
}
